File: module20, ensemble models/assign1.R

diabetes <- read.csv("C:/Users/USER/Downloads/Diabetes_RF.csv")

install.packages("caret")
library(caret)

#Accuracy with single model
inTraininglocal<-createDataPartition(diabetes$Class.variable,p=.75,list = F)
training<-diabetes[inTraininglocal,]
testing<-diabetes[-inTraininglocal,]

install.packages("C50")
library(C50)
model<-C5.0(training$Class.variable~.,data = training[,-9])
pred<-predict.C5.0(model,testing[,-9])
a<-table(testing$Class.variable,pred)
a
sum(diag(a))/sum(a) # accuracy is 69.79%

########Bagging
acc<-c()
for(i in 1:11)
{
  inTraininglocal<-createDataPartition(diabetes$Class.variable,p=.85,list = F)
  training1<-diabetes[inTraininglocal,]
  testing<-diabetes[-inTraininglocal,]
  fittree <- C5.0(training1$Class.variable~., data=training1[,-9])
  pred<-predict.C5.0(fittree,testing[,-9])
  a<-table(testing$Class.variable,pred)
  acc<-c(acc,sum(diag(a))/sum(a))
}
summary(acc)
mean(acc) # accuracy improved to  75%

####################### Boosting


#Accuracy with single model with Boosting

inTraininglocal<-createDataPartition(diabetes$Class.variable,p=.75,list = F)
training<-diabetes[inTraininglocal,]
testing<-diabetes[-inTraininglocal,]

model<-C5.0(training$Class.variable~.,data = training[,-9],trials=10)
pred<-predict.C5.0(model,testing[,-9])
a<-table(testing$Class.variable,pred)

sum(diag(a))/sum(a) # accuracy increased to 79.68%

######## Bagging and Boosting
acc<-c()
for(i in 1:11)
{
  
  inTraininglocal<-createDataPartition(diabetes$Class.variable,p=.85,list = F)
  training1<-diabetes[inTraininglocal,]
  testing<-diabetes[-inTraininglocal,]
  
  fittree <- C5.0(training1$Class.variable~., data=training1,trials=10)
  pred<-predict.C5.0(fittree,testing[,-9])
  a<-table(testing$Class.variable,pred)
  
  acc<-c(acc,sum(diag(a))/sum(a))
  
}

summary(acc)
mean(acc) # accuracy 73.83%

File: module20, ensemble models/assign1.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jun 13 22:46:12 2020

@author: USER
"""

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import VotingClassifier

data = pd.read_csv("C:/Users/USER/Downloads/Diabetes_RF.csv") #importing the dataset 
data.head() #seeting the head of the data

df_x = data.iloc[:,:8] #dividing the i/p and o/p variable
df_y = data.iloc[:,8]

x_train,x_test,y_train,y_test= train_test_split(df_x,df_y,test_size=0.2, random_state= 4) #dividing the data randomly
y_test.head()

#decision tree
dt = DecisionTreeClassifier() #storing the classifer in dt

dt.fit(x_train,y_train) #fitting te model 

dt.score(x_test,y_test) #checking the score like accuracy, 68.83%

dt.score(x_train,y_train) # 100% accuracy, difference is quite large between train and test accuracy, so there is problem of either overfitting or underfitting

#Random Forest clssifer: it is a ensemble of Decision tree 
rf = RandomForestClassifier(n_estimators=10) # n_estimator number of tree in the forest 
rf.fit(x_train,y_train) #fitting the random forest model 

rf.score(x_test,y_test) #doing the accuracy of the test model , 75.32% accuracy

rf.score(x_train,y_train) #doing the accuracy of the train model, 98.20%, slightly diff is reduced between train and test accuracy

#Bagging - Gradient 
bg = BaggingClassifier(DecisionTreeClassifier(), max_samples=0.5,max_features=1.0, n_estimators=20)
bg.fit(x_train,y_train) 

bg.score(x_test,y_test) #test accuracy, 75.97%

bg.score(x_train,y_train) #train accuracy , 95.11%, difference further reduced

#Ada boosting 
ada = AdaBoostClassifier(DecisionTreeClassifier(),n_estimators=10,learning_rate=1)
ada.fit(x_train,y_train)

ada.score(x_test,y_test) # 69.48% accuracy

ada.score(x_train,y_train) # 100% accuracy

# extreme gradient boosting

from xgboost import XGBClassifier
model = XGBClassifier()

model.fit(x_train,y_train)

model.score(x_test,y_test) # 75.32%

model.score(x_train,y_train) # 100%










